crates/evm-ethereum/src/execute.rs

```diff
@@ -0,0 +1,824 @@
+}
+
+impl EthExecutorProvider<EthEvmConfig> {
+    /// Creates a new default ethereum executor provider.
+    pub fn ethereum(chain_spec: Arc<ChainSpec>) -> Self {
+        Self::new(chain_spec, Default::default())
+    }
+}
+
+impl<EvmConfig> EthExecutorProvider<EvmConfig> {
+    /// Creates a new executor provider.
+    pub fn new(chain_spec: Arc<ChainSpec>, evm_config: EvmConfig) -> Self {
+        Self { chain_spec, evm_config, inspector: None, prune_modes: PruneModes::none() }
+    }
+
+    /// Configures an optional inspector stack for debugging.
+    pub fn with_inspector(mut self, inspector: InspectorStack) -> Self {
+        self.inspector = Some(inspector);
+        self
+    }
+
+    /// Configures the prune modes for the executor.
+    pub fn with_prune_modes(mut self, prune_modes: PruneModes) -> Self {
+        self.prune_modes = prune_modes;
+        self
+    }
+}
+
+impl<EvmConfig> EthExecutorProvider<EvmConfig>
+where
+    EvmConfig: ConfigureEvm,
+    EvmConfig: ConfigureEvmEnv<TxMeta = ()>,
+{
+    fn eth_executor<DB>(&self, db: DB) -> EthBlockExecutor<EvmConfig, DB>
+    where
+        DB: Database<Error = ProviderError>,
+    {
+        EthBlockExecutor::new(
+            self.chain_spec.clone(),
+            self.evm_config.clone(),
+            State::builder().with_database(db).with_bundle_update().without_state_clear().build(),
+        )
+        .with_inspector(self.inspector.clone())
+    }
+}
+
+impl<EvmConfig> ExecutorProvider for EthExecutorProvider<EvmConfig>
+where
+    EvmConfig: ConfigureEvm,
+    EvmConfig: ConfigureEvmEnv<TxMeta = ()>,
+{
+    type Executor<DB: Database<Error = ProviderError>> = EthBlockExecutor<EvmConfig, DB>;
+
+    type BatchExecutor<DB: Database<Error = ProviderError>> = EthBatchExecutor<EvmConfig, DB>;
+
+    fn executor<DB>(&self, db: DB) -> Self::Executor<DB>
+    where
+        DB: Database<Error = ProviderError>,
+    {
+        self.eth_executor(db)
+    }
+
+    fn batch_executor<DB>(&self, db: DB) -> Self::BatchExecutor<DB>
+    where
+        DB: Database<Error = ProviderError>,
+    {
+        let executor = self.eth_executor(db);
+        EthBatchExecutor {
+            executor,
+            batch_record: BlockBatchRecord::new(self.prune_modes.clone()),
+            stats: BlockExecutorStats::default(),
+        }
+    }
+}
+
+/// Helper container type for EVM with chain spec.
+#[derive(Debug, Clone)]
+struct EthEvmExecutor<EvmConfig> {
+    /// The chainspec
+    chain_spec: Arc<ChainSpec>,
+    /// How to create an EVM.
+    evm_config: EvmConfig,
+}
+
+impl<EvmConfig> EthEvmExecutor<EvmConfig>
+where
+    EvmConfig: ConfigureEvm,
+    EvmConfig: ConfigureEvmEnv<TxMeta = ()>,
+{
+    /// Executes the transactions in the block and returns the receipts.
+    ///
+    /// This applies the pre-execution changes, and executes the transactions.
+    ///
+    /// # Note
+    ///
+    /// It does __not__ apply post-execution changes.
+    fn execute_pre_and_transactions<Ext, DB>(
```

Comment:
maybe nice to not bundle pre-execute and execute in one call, since it already has this nice division of logic